### PR TITLE
Fix WebException to handle SocketsHttpHandler generated errors

### LINKED ISSUE #27906: GetResponseAsync_ServerNameNotInDns_ThrowsWebException() test fails with SocketsHttpHandler
Expecting `WebExceptionStatus.NameResolutionFailure` get `UnknownError` instead.

Call stack for WinHttpHandler:
```

+		ex	{System.Net.WebException: An error occurred while sending the request. The server name or address could not be resolved ---> System.Net.Http.HttpRequestException: An error occurred while sending the request. ---> System.Net.Http.WinHttpException: The server name or address could not be resolved
   at System.Environment.get_StackTrace() in D:\corefx\src\System.Runtime.Extensions\src\System\Environment.cs:line 146
   at System.Runtime.ExceptionServices.ExceptionStackTrace.AddCurrentStack(Exception exception) in D:\corefx\src\Common\src\System\Runtime\ExceptionServices\ExceptionStackTrace.cs:line 23
   at System.Net.Http.WinHttpException.CreateExceptionUsingError(Int32 error) in D:\corefx\src\Common\src\System\Net\Http\WinHttpException.cs:line 56
   at System.Net.Http.WinHttpRequestCallback.OnRequestError(WinHttpRequestState state, WINHTTP_ASYNC_RESULT asyncResult) in D:\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs:line 324
   at System.Net.Http.WinHttpRequestCallback.RequestCallback(IntPtr handle, WinHttpRequestState state, UInt32 internetStatus, IntPtr statusInformation, UInt32 statusInformationLength) in D:\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs:line 104
   at System.Net.Http.WinHttpRequestCallback.WinHttpCallback(IntPtr handle, IntPtr context, UInt32 internetStatus, IntPtr statusInformation, UInt32 statusInformationLength) in D:\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpRequestCallback.cs:line 47
--- End of stack trace from AddCurrentStack ---
   at System.Threading.Tasks.RendezvousAwaitable`1.GetResult() in D:\corefx\src\Common\src\System\Threading\Tasks\RendezvousAwaitable.cs:line 62
   at System.Net.Http.WinHttpHandler.StartRequest(WinHttpRequestState state) in D:\corefx\src\System.Net.Http.WinHttpHandler\src\System\Net\Http\WinHttpHandler.cs:line 856
   --- End of inner exception stack trace ---
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\DiagnosticsHandler.cs:line 98
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in D:\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 501
   at System.Net.HttpWebRequest.SendRequest() in D:\corefx\src\System.Net.Requests\src\System\Net\HttpWebRequest.cs:line 1209
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult) in D:\corefx\src\System.Net.Requests\src\System\Net\HttpWebRequest.cs:line 1259
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult) in D:\corefx\src\System.Net.Requests\src\System\Net\HttpWebRequest.cs:line 1263
   at System.Net.WebRequest.<>c.<GetResponseAsync>b__68_2(IAsyncResult iar) in D:\corefx\src\System.Net.Requests\src\System\Net\WebRequest.cs:line 550
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization) in E:\A\_work\1045\s\src\mscorlib\src\System\Threading\Tasks\FutureFactory.cs:line 529
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Assert.RecordExceptionAsync(Func`1 testCode) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.assert\Asserts\Record.cs:line 76}	System.Net.WebException



		ex.Status	NameResolutionFailure	System.Net.WebExceptionStatus
```

Call stack for SocketsHttpHandler:
```
+		ex	{System.Net.WebException: No such host is known No such host is known ---> System.Net.Http.HttpRequestException: No such host is known ---> System.Net.Sockets.SocketException: No such host is known
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\ConnectHelper.cs:line 78
   --- End of inner exception stack trace ---
   at System.Net.Http.ConnectHelper.ConnectAsync(String host, Int32 port, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\ConnectHelper.cs:line 98
   at System.Threading.Tasks.ValueTask`1.get_Result() in E:\A\_work\1045\s\src\mscorlib\shared\System\Threading\Tasks\ValueTask.cs:line 713
   at System.Net.Http.HttpConnectionPool.CreateConnectionAsync(HttpRequestMessage request, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs:line 309
   at System.Threading.Tasks.ValueTask`1.get_Result() in E:\A\_work\1045\s\src\mscorlib\shared\System\Threading\Tasks\ValueTask.cs:line 713
   at System.Net.Http.HttpConnectionPool.WaitForCreatedConnectionAsync(ValueTask`1 creationTask) in D:\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs:line 450
   at System.Threading.Tasks.ValueTask`1.get_Result() in E:\A\_work\1045\s\src\mscorlib\shared\System\Threading\Tasks\ValueTask.cs:line 713
   at System.Net.Http.HttpConnectionPool.SendWithRetryAsync(HttpRequestMessage request, Boolean doRequestAuth, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\HttpConnectionPool.cs:line 243
   at System.Net.Http.RedirectHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\SocketsHttpHandler\RedirectHandler.cs:line 30
   at System.Net.Http.DiagnosticsHandler.SendAsync(HttpRequestMessage request, CancellationToken cancellationToken) in D:\corefx\src\System.Net.Http\src\System\Net\Http\DiagnosticsHandler.cs:line 98
   at System.Net.Http.HttpClient.FinishSendAsyncUnbuffered(Task`1 sendTask, HttpRequestMessage request, CancellationTokenSource cts, Boolean disposeCts) in D:\corefx\src\System.Net.Http\src\System\Net\Http\HttpClient.cs:line 501
   at System.Net.HttpWebRequest.SendRequest() in D:\corefx\src\System.Net.Requests\src\System\Net\HttpWebRequest.cs:line 1209
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult) in D:\corefx\src\System.Net.Requests\src\System\Net\HttpWebRequest.cs:line 1259
   --- End of inner exception stack trace ---
   at System.Net.HttpWebRequest.EndGetResponse(IAsyncResult asyncResult) in D:\corefx\src\System.Net.Requests\src\System\Net\HttpWebRequest.cs:line 1263
   at System.Net.WebRequest.<>c.<GetResponseAsync>b__68_2(IAsyncResult iar) in D:\corefx\src\System.Net.Requests\src\System\Net\WebRequest.cs:line 550
   at System.Threading.Tasks.TaskFactory`1.FromAsyncCoreLogic(IAsyncResult iar, Func`2 endFunction, Action`1 endAction, Task`1 promise, Boolean requiresSynchronization) in E:\A\_work\1045\s\src\mscorlib\src\System\Threading\Tasks\FutureFactory.cs:line 529
--- End of stack trace from previous location where exception was thrown ---
   at Xunit.Assert.RecordExceptionAsync(Func`1 testCode) in C:\BuildAgent\work\cb37e9acf085d108\src\xunit.assert\Asserts\Record.cs:line 76}	System.Net.WebException


		ex.Status	UnknownError	System.Net.WebExceptionStatus
```

### PR DESCRIPTION
**Pull request by davidsh**
Now that we are using the new handler within HttpClient, we have to revise the code
that maps various exceptions to enum values of the WebExceptionStatus. We have separate
versions of this mapping for Windows vs. Linux. But the new handler runs on both. So,
added a common helper method.

There will probably be additional mappings we need to do here. But for now, this
PR fixes the issue.

Fixes #27906

### CODE REVIEW

**Comment by davidsh** [comment]
cc: @dotnet/ncl

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

File: src/System.Net.Requests/src/System/Net/WebException.cs [hunk 1]
```diff
@@ -110,5 +111,28 @@ internal static Exception CreateCompatibleException(Exception exception)
 
             return exception;
         }
+        
+        private static WebExceptionStatus GetStatusFromExceptionHelper(HttpRequestException ex)
+        {
+            WebExceptionStatus status;
```
Comment (stephentoub) [inline src/System.Net.Requests/src/System/Net/WebException.cs]: Nit: this can be moved down to just before the switch

Comment (davidsh) [reply src/System.Net.Requests/src/System/Net/WebException.cs]: Fixed

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

**Comment by davidsh** [comment]
This is failing on Linux on the re-activated test.  The errors are:

```
Assert.Equal() Failure
Expected: NameResolutionFailure
Actual:   UnknownError
```

This implies that the underlying Socket error is different on Linux than Windows. On Windows, the DNS resolution failures generate `SocketError.HostNotFound`.  Not sure why this would be different on Linux. There might be an underlying Socket bug regarding translating the Linux errors to `SocketError` enum values.  Need to investigate.

cc: @geoffkizer @wfurt

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build

**Comment by wfurt** [comment]
This could be further OS dependend @davidsh I've seen exception "device not found" on name resolution failure. The real reason may be deep down in curl.

**Comment by davidsh** [comment]
>The real reason may be deep down in curl.

This is not using Curl anymore.  HttpWebRequest always uses the default handler for HttpClient.  That default is now SocketsHttpHandler.  So, both Windows and Linux are using the new handler only.  So, it is directly using Socket and not Curl.

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Linux x64 Debug Build

**Comment by davidsh** [comment]
It looks like the fix to this is to use SocketException.SocketErrorCode and not SocketException.ErrorCode property.  The `ErrorCode` property is always the native error value. On Windows this is the same as the `SocketError` enum values. But on Linux, it is Linux native error values. On Windows, DNS resolution errors return, 11001 for "host not found". But on Linux. it is a value of 6 for "No such device or address".

The `SocketException.SocketErrorCode` property is always one of the enum values. So, the translation of native error codes is done in the PAL layer of System.Net.Sockets. So, I fixed my PR to use SocketErrorCode property and now the test passes on both Windows and Linux.

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

**Comment by davidsh** [comment]
@dotnet-bot test Outerloop Windows x64 Debug Build
@dotnet-bot test Outerloop Linux x64 Debug Build

**Comment by davidsh** [comment]
CI failures not related to PR.

**Pull request merged**